pandas/core/indexes/multi.py

```diff
@@ -1412,6 +1420,107 @@ def from_product(cls, iterables, sortorder=None, names=None):
+
+        Returns
+        -------
+        MultiIndex or Index
+            The MultiIndex representation of the given DataFrame. Returns an
+            Index if the DataFrame is single column and squeeze is True.
+
+        Examples
+        --------
+        >>> df = pd.DataFrame([[0, 'happy'], [0, 'jolly'], [1, 'happy'],
+        ...                    [1, 'jolly'], [2, 'joy'], [2, 'joy']],
+        ...                   columns=['number', 'mood'])
+        >>> df
+           number   mood
+        0       0  happy
+        1       0  jolly
+        2       1  happy
+        3       1  jolly
+        4       2    joy
+        5       2    joy
+        >>> pd.MultiIndex.from_frame(df)
+        MultiIndex(levels=[[0, 1, 2], ['happy', 'jolly', 'joy']],
+                   labels=[[0, 0, 1, 1, 2, 2], [0, 1, 0, 1, 2, 2]],
+                   names=['number', 'mood'])
+        >>> df = pd.DataFrame([[0], [1]], columns=['number'])
+        >>> df
+           number
+        0       0
+        1       1
+        >>> pd.MultiIndex.from_frame(df)
+        Int64Index([0, 1], dtype='int64', name='number')
+        >>> pd.MultiIndex.from_frame(df, squeeze=False)
+        MultiIndex(levels=[[0, 1]],
+                   labels=[[0, 1]],
+                   names=['number'])
+        >>> df = pd.DataFrame([['a', 'b'], ['c', 'd'], ['e', 'f']],
+        ...                   columns=pd.MultiIndex.from_tuples([('L1', 'x'),
+        ...                                                      ('L2', 'y')]))
+        >>> df
+          L1 L2
+           x  y
+        0  a  b
+        1  c  d
+        2  e  f
+        >>> pd.MultiIndex.from_frame(df, names=lambda x: '_'.join(x))
+        MultiIndex(levels=[['a', 'c', 'e'], ['b', 'd', 'f']],
+                   labels=[[0, 1, 2], [0, 1, 2]],
+                   names=['L1_x', 'L2_y'])
+
+        See Also
+        --------
+        MultiIndex.from_arrays : Convert list of arrays to MultiIndex
+        MultiIndex.from_tuples : Convert list of tuples to MultiIndex
+        MultiIndex.from_product : Make a MultiIndex from cartesian product
+                                  of iterables
+        """
+        from pandas import DataFrame
+        if not isinstance(df, DataFrame):
+            raise TypeError("Input must be a DataFrame")
+
+        # Get MultiIndex names
+        if names is None:
+            names = list(df)
+        elif callable(names):
+            names = [names(x) for x in list(df)]
+        elif is_list_like(names):
+            if len(names) != len(list(df)):
+                    raise ValueError("'names' should have same length as "
+                                     "number of columns in df.")
+            # else: use the passed in sequence
+        else:
+            raise TypeError("'names' must be a list / sequence of column "
+                            "names, or a callable.")
+
+        # This way will preserve dtype of columns
+        mi = cls.from_arrays([df[x] for x in df],
```

Comment:
I would raise on ``not df.columns.is_unique``

Reply:
since non-unique names are valid in multiindex I have updated from_frame to provide the analogous multiindex even in the case of repeat names (like so: `[df.iloc[:, x] for x in range(len(df.columns))]`